### PR TITLE
Add the first test for the agent

### PR DESCRIPTION
**Pull request by kevinearls**
Signed-off-by: Kevin Earls <kearls@redhat.com>

### CODE REVIEW

**Comment by kevinearls** [comment]
@jpkrohling  @objectiser  Please review

File: test/e2e/config/jaeger-collector-config.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+extensions:
```
Comment (objectiser) [inline test/e2e/config/jaeger-collector-config.yaml:1]: Is this file required for the agent e2e tests?

File: test/e2e/utils.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+			return m
+		}
+	}
+
+	logrus.Warnf("Could not find metric %s at endpoint %s", metricKey, metricsEndpoint)
+	emptyMetric := Metric{}
+	return emptyMetric
+}
+
+func getMetrics(t *testing.T, metricsEndpoint string) []Metric {
```
Comment (objectiser) [inline test/e2e/utils.go:60]: Not necessarily a big issue now, but wondering whether there is a standard client lib we could use to retrieve the metrics rather than parsing ourselves? If so, then maybe create an issue to track migrating this to a standard lib.

**Review by objectiser — COMMENTED** [review]
Generally looks good to me - although if possible I would prefer the tests to be co-located with the agent module, with only common utilities perhaps at the top level?

Although not so much of an issue for the agent, some components may have dependencies only relevant for testing individual components, so could be good to keep them local.

@jpkrohling thoughts?

File: test/e2e/agent_sanity_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	agent := StartCollector(t, agentExecutable, agentConfigFileName, &agentLoggerOutput, metricsPort)
+	defer agent.Process.Kill()
+
+	// Create some traces. Each trace created by tracegen will have 2 spans
+	traceCount := 5
+	expectedSpanCount := 2 * traceCount
+	serviceName := "agent-sanity-test" + strconv.Itoa(time.Now().Nanosecond())
+	tracegen.CreateJaegerTraces(t, 1, traceCount, 0, serviceName)
+
+	// This could be changed to logrus.Debugf if we can stop logrus from eating newlines
```
Comment (jpkrohling) [inline test/e2e/agent_sanity_test.go:61]: Isn't it better then to store this output in another file, and display the contents of this file in case the test fails?

File: test/e2e/agent_sanity_test.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	if logrus.GetLevel() == logrus.DebugLevel {
+		fmt.Printf("%s", agentLoggerOutput.String())
+	}
+
+	// Check the agent output for service name and Span entries
+	require.Contains(t, agentLoggerOutput.String(), "service.name: STRING("+serviceName+")")
+
+	spanExpression := regexp.MustCompile("Span #")
```
Comment (jpkrohling) [inline test/e2e/agent_sanity_test.go:69]: I see now what it is doing... You are relying on the output of the loggingexporter for this test. Isn't it possible to get the metrics for the receiver/processor/exporter instead? If it fails, you could log the logs for the three phases, so that we know where the problem might be.

File: test/e2e/agent_sanity_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	index := suffixarray.New(agentLoggerOutput.Bytes())
+	results := index.FindAllIndex(spanExpression, -1)
+	require.Equal(t, expectedSpanCount, len(results))
+
+	// Check the metrics to verify that the agent received and then sent the number of spans expected
+	metricsEndpoint := "http://localhost:" + metricsPort + "/metrics"
+	receivedSpansMetric := getMetric(t, metricsEndpoint, "otelcol_receiver_accepted_spans")
+	sentSpansMetric := getMetric(t, metricsEndpoint, "otelcol_exporter_sent_spans")
+	require.Equal(t, strconv.Itoa(expectedSpanCount), receivedSpansMetric.Value)
+	require.Equal(t, strconv.Itoa(expectedSpanCount), sentSpansMetric.Value)
```
Comment (jpkrohling) [inline test/e2e/agent_sanity_test.go:79]: Oh, you *are* checking the metrics. Then why do you need to parse the logs?

File: test/e2e/utils.go (hunk 2 above)
Comment (kevinearls) [reply test/e2e/utils.go:60]: I've added https://github.com/jaegertracing/jaeger-otelcol/issues/15

File: test/e2e/agent_sanity_test.go (hunk 5 above)
Comment (kevinearls) [reply test/e2e/agent_sanity_test.go:79]: I suppose I don't really *need* to, it's just a bit of a sanity check.  I can remove it.

**Comment by kevinearls** [comment]
@objectiser @jpkrohling Aren't you guys supposed to be on PTO? 😁

**Comment by objectiser** [comment]
@kevinearls Just a few observations to help things along :smile: 

@jpkrohling Thoughts on whether the agent e2e tests should be moved to its module (i.e. more of a modular approach)? Rather than having all e2e tests together?

**Comment by jpkrohling** [comment]
> Thoughts on whether the agent e2e tests should be moved to its module  

Sounds like a good idea.

**Comment by kevinearls** [comment]
@objectiser @jpkrohling Sorry if this is a dumb question, but as far as moving the agent e2e tests to its module, where exactly would that be?

**Comment by objectiser** [comment]
@kevinearls sorry ignore me - forgot that part is auto-generated :)

However might be good to have sub-folders under the `test/e2e` for each component's specific tests?

File: Makefile [hunk 6]
```diff
@@ -31,6 +31,14 @@ otelcol-builder:
 ifeq (, $(shell which opentelemetry-collector-builder))
 	go get github.com/observatorium/opentelemetry-collector-builder@v$(OTELCOL_BUILDER_VERSION)
 endif
+
+.PHONY: e2e-tests
+e2e-tests: build e2e-tests-agent-smoke
+
+.PHONY: e2e-tests-agent-smoke
+e2e-tests-agent-smoke: build-agent
+	@echo Running Agent  end-to-end tests...
```
Comment (objectiser) [inline Makefile:40]: nit: remove space after Agent

**Review by objectiser — COMMENTED** [review]
Only a minor comment. But @jpkrohling had some comments - not sure if addressed? https://github.com/jaegertracing/jaeger-otelcol/pull/14#discussion_r546932785 and https://github.com/jaegertracing/jaeger-otelcol/pull/14#discussion_r546933845

**Review by objectiser — APPROVED** [review]
I'll approve/merge - if Juca's comments are still relevant then may need to be handled in a separate PR.

**Comment by objectiser** [comment]
Merging manually as:
"Pull request must be merged manually.
GitHub App like Mergify are not allowed to merge pull request where .github/workflows is changed."

File: test/e2e/agent/agent_sanity_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
+// +build agent_smoke
+
+package e2e
+
+import (
+	"fmt"
+	"github.com/jaegertracing/jaeger-otelcol/test/e2e"
```
Comment (jpkrohling) [inline test/e2e/agent/agent_sanity_test.go:7]: This repo might need a linter to make sure imports are in the right order.

File: test/e2e/agent/agent_sanity_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	// Create some traces. Each trace created by tracegen will have 2 spans
+	traceCount := 5
+	expectedSpanCount := 2 * traceCount
+	serviceName := "agent-sanity-test" + strconv.Itoa(time.Now().Nanosecond())
+	tracegen.CreateJaegerTraces(t, 1, traceCount, 0, serviceName)
+
+	// This could be changed to logrus.Debugf if we can stop logrus from eating newlines
+	if logrus.GetLevel() == logrus.DebugLevel {
+		log, err := ioutil.ReadFile(loggerOutputFile.Name())
```
Comment (jpkrohling) [inline test/e2e/agent/agent_sanity_test.go:66]: I still think it would be better to not touch the logs here. If the test fails, `cat` the log file to the CI output instead. When running locally, I can cat/grep the logs myself as well.
